train.py

```diff
@@ -0,0 +1,319 @@
+            top_p: Optional[torch.Tensor] = None,
+            top_k: Optional[torch.Tensor] = None,
+            max_new_tokens: Optional[int] = None,
+            end_of_audio_token: int = 2048,
+            device: str = "cuda",
+    ) -> torch.Tensor:
+        # Should return decoded Encodec tokens from stage 1 LLM as a torch.Tensor
+        T = prompt.size(0)
+        if max_new_tokens is None:
+                max_seq_length = self.model.config.block_size
+        else:
+            max_seq_length = T + max_new_tokens
+            max_seq_length = min(max_seq_length, self.model.config.block_size)
+        max_new_tokens = max_seq_length - T
+        if max_new_tokens <= 0:
+            raise ValueError("Prompt is too long to generate more tokens")
+    
+        device, dtype = prompt.device, prompt.dtype
+
+        seq = torch.clone(prompt)
+        input_pos = torch.arange(0, T, device=device)
+
+        next_token: torch.Tensor = prefill(
+            self.model, 
+            prompt.view(1, -1).repeat(2, 1), 
+            spk_emb, 
+            input_pos, 
+            guidance_scale = guidance_scale, 
+            temperature = temperature, 
+            top_p = top_p, 
+            top_k = top_k
+        )
+        seq = torch.cat([seq, next_token.view(1)])
+
+        input_pos = torch.tensor([T], device=device, dtype=torch.int)
+
+        generated_tokens, _ = decode_n_tokens(
+            self.model,
+            next_token.view(1, -1).repeat(2, 1),
+            spk_emb,
+            input_pos,
+            max_new_tokens - 1,
+            callback=lambda x: x,
+            end_of_audio_token=end_of_audio_token,
+            guidance_scale=guidance_scale,
+            temperature=temperature,
+            top_p=top_p,
+            top_k=top_k,
+        )
+
+        seq = torch.cat([seq, torch.cat(generated_tokens)])
+
+        tokens = seq.tolist()
+
+        _, extracted_audio_ids = self.first_stage_adapter.decode([tokens])
+
+        return [torch.tensor(extracted_audio_ids, dtype=torch.int32, device=self._device).unsqueeze(0)]
+
+    def train(self, training_name: str, epochs=10, learning_rate=1e-4):
+        print("Initializing dataloader...")
+        print("SECOND STAGE CAUSAL:")
+        print(self.llm_second_stage.model.config.causal)
+
+        # Model configuration for training
+        top_p = 0.95
+        guidance_scale = 3.0
+        temperature = 1.0
+
+        dataloader = create_metavoice_dataloader(
+            dataset_dir=self._dataset_dir,
+            encodec_model=self.encodec_model,
+            tokenizer=self.tokenizer, 
+            spkemb_model=self.smodel, 
+            batch_size=2, 
+            device=self._device)
+        # TODO(validation data) load validation data as well for val loss
+
+        print("Setting model to training mode...")
+        self.model.train()  # Set the model to training mode
+        self.llm_second_stage.model.train()  # Set the model to training mode
```

Comment:
We probably don't need to train this stage at all if you're only adapting speaker identities

Reply:
Agree. Removed stage 2 in latest commit :-)